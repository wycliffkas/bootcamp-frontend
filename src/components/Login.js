import React, { Component } from 'react';
import { connect } from 'react-redux';
import { loginUser } from '../actions/questions';

class  Login extends Component {
    constructor(props){
        super(props);
        this.state = {
            username: '',
            password: ''
        }
        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onChange(e){
        this.setState({[e.target.name]:e.target.value});
    }

    onSubmit(e){
        e.preventDefault();
        const user = {
            username: this.state.username,
            password: this.state.password
        };

        this.props.loginUser(user);
    }


    render(){
        return (
            <div className="login-form">
                <form onSubmit={this.onSubmit}>
                <div className="top">
                    <h3>LOGIN</h3>
                </div>
                <div className="form-area">
                    <div className="group">
                    <input type="text" className="form-control"  name="username"  value={this.state.username} onChange={this.onChange} placeholder="Username"/>
                    <i className="fa fa-user"></i>
                    </div>
                    <div className="group">
                    <input type="password" className="form-control" name="password" value={this.state.password} onChange={this.onChange} placeholder="Password"/>
                    <i className="fa fa-key"></i>
                    </div>
                    <button type="submit" className="btn btn-default btn-block">Submit</button>
                    <div className="checkbox checkbox-primary">
                    <br/>
                    <a href="/register">Register Now</a>
                    </div>                    
                </div>
                </form>
            </div>
        );
    };
}

export const mapDispatchToProps = dispatch => ({
    loginUser: data => dispatch(loginUser(data))
  });
  export { Login };
  
  export default connect(
    null,
    mapDispatchToProps
  )(Login);
