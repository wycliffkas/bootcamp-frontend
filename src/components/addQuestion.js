import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addNewQuestion } from '../actions/questions';

class AddQuestion extends Component {

    constructor(props){
       super(props);
       this.state = {
           title: '',
           description: ''
       }

       this.onChange = this.onChange.bind(this);
       this.onSubmit = this.onSubmit.bind(this);
       
    }

    onChange(e){
        this.setState({[e.target.name]:e.target.value});
    }

    onSubmit(e){
        e.preventDefault();
        const question = {
            question: this.state.title,
            description: this.state.description,
        };

        this.props.addNewQuestion(question);

    }   

    render(){
        return (
            <div className="login-form">
                <form onSubmit={this.onSubmit}>
                <div className="top">
                    <h3>Add a Question</h3>
                </div>
                <div className="form-area">
                    <div className="group">
                    Title
                    <input type="text" className="form-control"  name="title"  value={this.state.title} onChange={this.onChange} required/>
                    </div>
                    <div className="group">
                    Description
                    <textarea className="form-control-other" rows="3" name="description" value={this.state.description} onChange={this.onChange} required></textarea>
                    </div>                                      
                    <button type="submit" className="btn btn-default btn-block">Submit</button>                  
                </div>
                </form>
            </div>
        );

    };
}


export const mapStateToProps = state => ({
    question: state.questions.question
  });
  
  export const mapDispatchToProps = dispatch => ({
    addNewQuestion: data => dispatch(addNewQuestion(data))
  });
  export { AddQuestion };
  
  export default connect(
    mapStateToProps,
    mapDispatchToProps
  )(AddQuestion);