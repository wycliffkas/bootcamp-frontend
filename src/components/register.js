import React, { Component } from 'react';
import { connect } from 'react-redux';
import { registerUser } from '../actions/questions';

class  Register extends Component {

   constructor(props){
       super(props);
       this.state = {
           fullname: '',
           username: '',
           email: '',
           password: ''
       }
       this.onChange = this.onChange.bind(this);
       this.onSubmit = this.onSubmit.bind(this);
    }

    onChange(e){
        this.setState({[e.target.name]:e.target.value});
    }

    onSubmit(e){
        e.preventDefault();
        const user = {
            fullname: this.state.fullname,
            username: this.state.username,
            email: this.state.email,
            password: this.state.password
        };

        this.props.registerUser(user);

    }


    render() {
        return (
            <div className="login-form">
                <form onSubmit={this.onSubmit}>
                <div className="top">
                    <h3>Register User</h3>
                </div>
                <div className="form-area">
                    <div className="group">
                    <input type="text" className="form-control" name="fullname" value={this.state.fullname} onChange={this.onChange} placeholder="Names" required/>
                    </div>                
                    <div className="group">
                    <input type="text" className="form-control" name="username" value={this.state.username} onChange={this.onChange} placeholder="Username" required/>
                    </div>
                    <div className="group">
                    <input type="text" className="form-control" name="email" value={this.state.email} onChange={this.onChange} placeholder="Email" required/>
                    </div>                    
                    <div className="group">
                    <input type="password" className="form-control" name="password" value={this.state.password} onChange={this.onChange} placeholder="Password" required/>
                    </div>                  
                    <button type="submit" className="btn btn-default btn-block">Save</button>                  
                </div>

                </form>
            </div>
        );
    };
}

// export default connect(null, { registerUser })(Register);
const mapStateToProps = state => ({
    question: state
  });
  
  export const mapDispatchToProps = dispatch => ({
    register: data => dispatch(register(data))
  });
  
  export { Register };
  
  export default connect(
    mapStateToProps,
    mapDispatchToProps
  )(Register);

