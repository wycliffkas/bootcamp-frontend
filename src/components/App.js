import React from 'react';
import { Route, Switch } from 'react-router-dom';
import Header from './pages/Header';
import Footer from './pages/Footer';
import GetAllQuestions from './GetAllQuestions';
import GetSingleQuestion from './GetSingleQuestion';
import Register from './Register';
import Login from './Login';

const App = () => (
    <div>
        <Header /> 
        <div className="container">
            <Switch>
                <Route exact path="/" component={GetAllQuestions} />
                <Route exact path="/questions/:id" component={GetSingleQuestion} /> 
                <Route exact  path="/register" component={Register} />  
                <Route exact path="/login" component={Login} />
            </Switch>
        </div>
        <Footer />
    </div>
);

export default App;
