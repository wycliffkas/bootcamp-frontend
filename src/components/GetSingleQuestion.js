import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchQuestion } from '../actions/questions';
import "../App.css";

class GetSingleQuestion extends Component {

    componentWillMount(){
        const { match, fetchQuestion} = this.props;
        const { id } = match.params;
        fetchQuestion(id);
    }


    render() {
        const { question } = this.props;
        const questionItem = question => (
            <div className="panel panel-default">
                    <h4>{question.question}</h4>
                <p>{question.description}</p>
                <b>Author:</b> {question.author}
            </div>
        );

        return (
            <div>
                <h3>Question</h3> 
                <div>{ questionItem(question) }</div>
            </div>
        )        

    }
}

const mapStateToProps = state => ({
    question: state.questions.question,
});

export const mapDispatchToProps = dispatch => ({
    fetchQuestion: (id) => dispatch(fetchQuestion(id))
});

export { GetSingleQuestion };
export default connect(mapStateToProps, mapDispatchToProps)(GetSingleQuestion);