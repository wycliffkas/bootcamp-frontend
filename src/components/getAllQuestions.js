import React, { Component } from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import PropTypes from "prop-types";
import { fetchQuestions } from '../actions/questions';
import "../App.css";

class GetAllQuestions extends Component {

    componentWillMount(){
        this.props.fetchQuestions();
    }

    
    render() {
        const { questions } = this.props;
        const questionsItems = (questions).map((question) => (
            <div key={question.questionId} className="panel panel-default">
                    
                <NavLink to={{pathname: `/questions/${question.questionId}`}}>
                    <h4>{question.question}</h4>
                </NavLink>
                <p>{question.description}</p>
                <b>Author:</b> {question.author}
            </div>
        ));
        return (
            <div>
                <div>
                    {questions.length > 1 ? 
                    (
                        <div>
                    <h3>All Questions</h3> 
                    <div>{questionsItems}</div>
                    </div>
                    )
                    : <div className="text-center">
                    <img
                      src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
                      className="loading__image"
                      alt=""
                    />
                  </div> }
                </div>
               
                
            </div>
        )
        

    }
}


export const mapStateToProps = state => ({
    questions: state.questions.questions,
});

export const mapDispatchToProps = dispatch => ({
    fetchQuestions: () => dispatch(fetchQuestions())
});

export { GetAllQuestions };
export default connect(mapStateToProps, mapDispatchToProps)(GetAllQuestions);