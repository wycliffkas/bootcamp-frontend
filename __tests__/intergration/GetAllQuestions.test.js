import React from "react";
import { shallow, mount } from "enzyme";
import  { GetAllQuestions, mapDispatchToProps } from "../../src/components/GetAllQuestions";

describe("GetAllQuestions", () => {

	const initialState = {
	questions:[], 
	};

	const editor = shallow(<GetAllQuestions  questions={[initialState]} fetchQuestions={jest.fn()} />);
  
	it("should render without crashing", () => {
	  expect(editor).toMatchSnapshot();
	});
  
	it("it triggers fetchQuestions", () => {
	  const mock = jest.fn();
	  const mapper = mapDispatchToProps(mock);
	  mapper.fetchQuestions();
	  expect(mock).toHaveBeenCalled();
	});
});
