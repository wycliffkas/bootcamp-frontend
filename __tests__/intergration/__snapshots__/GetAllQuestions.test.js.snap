// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAllQuestions should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GetAllQuestions
    fetchQuestions={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    questions={
      Array [
        Object {
          "questions": Array [],
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          className="text-center"
        >
          <img
            alt=""
            className="loading__image"
            src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
          />
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="text-center"
        >
          <img
            alt=""
            className="loading__image"
            src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
          />
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            alt=""
            className="loading__image"
            src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
          />,
          "className": "text-center",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "",
            "className": "loading__image",
            "src": "https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            className="text-center"
          >
            <img
              alt=""
              className="loading__image"
              src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
            />
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="text-center"
          >
            <img
              alt=""
              className="loading__image"
              src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
            />
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt=""
              className="loading__image"
              src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
            />,
            "className": "text-center",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "",
              "className": "loading__image",
              "src": "https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
